adding three fields to clockwork.toml

gyom opened the pull request:
These fields are going to be used for calling sacct through SSH. These fields are already in the `hosts.yml` but they don't make it to `/etc/clockwork/clockwork.toml` at the moment and that means that we cannot use them. This commit will fix that.

This related to `cw204` branch of Clockwork, and in turn to `https://github.com/mila-iqia/clockwork/pull/88`.

gyom committed:
Adding three fields to clockwork.toml

These fields are going to be used for calling `sacct` through SSH. These fields are already in the `hosts.yml` but they don't make it to `/etc/clockwork/clockwork.toml` at the moment and that means that we cannot use them. This commit will fix that.

This related to cw204 branch of Clockwork, and in turn to `https://github.com/mila-iqia/clockwork/pull/88`.

gyom committed:
Merge pull request #1 from gyom/adding_fields_for_sacct

Adding three fields to clockwork.toml

abergeron commented on templates/clockwork.toml.j2:
```diff
@@ -35,3 +35,23 @@ allocations={{ cluster.allocations }}
+{% if cluster.remote_hostname %}
+remote_hostname="{{ cluster.remote_hostname }}"
+{% else %}
+remote_hostname=""
```
remote_hostname is there twice

gyom replied:
Removed.

abergeron commented on templates/clockwork.toml.j2:
```diff
@@ -35,3 +35,23 @@ allocations={{ cluster.allocations }}
 {% endif %}
 timezone="{{ cluster.timezone }}"
 {% endfor %}
```
All of your additions need to be inside this for loop otherwise it will not render correctly.

gyom replied:
Done.

abergeron commented on templates/clockwork.toml.j2:
```diff
@@ -35,3 +35,23 @@ allocations={{ cluster.allocations }}
+remote_user="{{ cluster.remote_user }}"
+{% else %}
+remote_user=""
+{% endif %}
```
This could be simplified as

```
remote_user="{{ cluster.remote_user }}"
```

without the if dance around it.  Same for remote_hostname

gyom replied:
Done.

abergeron commented on templates/clockwork.toml.j2:
```diff
@@ -35,3 +35,23 @@ allocations={{ cluster.allocations }}
+remote_hostname=""
+{% endif %}
+{% if cluster.sacct_enabled %}
+sacct_enabled="{{ cluster.sacct_enabled }}"
```
Is this a string of a boolean?  If it is a string, it should have quotes around the `"false"` below, if it is a boolean it should not have quotes here.

gyom replied:
Done.

abergeron reviewed (changes requested):
I'm wondering why you need to add those fields to the clockwork config.

Normally this kind of thing should be handled by the remote data script like it is currently done for the sinfo files.

If you want to rework this, I'm ok with it.

abergeron commented:
I found the PR is is https://github.com/mila-iqia/clockwork/pull/88

abergeron commented:
From reading the associated PR, there should also be a `sacct_path` field for each cluster.

gyom committed:
fixes after Arnaud's initial comments

Properly enclose things in the for loop.
Remove quotes where dealing with booleans.
Remove if/else when dealing with strings whose default values are "".

gyom commented:
> I'm wondering why you need to add those fields to the clockwork config.
> 
> Normally this kind of thing should be handled by the remote data script like it is currently done for the sinfo files.
> 
> If you want to rework this, I'm ok with it.

Is this request still relevant given the discussion yesterday and the changes today?

gyom committed:
added field for sacct ssh key filename

This might differ from cluster to cluster (even though right now it doesn't), so let's assign it to each cluster.

gyom committed:
adding more fields to sync with test_config.toml

Many fields were being used in dev with `test_config.toml`, but they hadn't made their way into this `clockwork.toml`.

abergeron commented on templates/clockwork.toml.j2:
```diff
@@ -34,4 +34,20 @@ allocations="*"
+sacct_enabled={{ cluster.sacct_enabled }}
+{% else %}
+sacct_enabled=false
+{% endif %}
```
This doesn't need to be in an if/else construct.

gyom replied:
Doesn't it?

If the this field is missing, won't the template will just leave it empty so it'll be
```
sacct_enabled=
```
which is going to cause issues with the parsing of the configuration (I'm guessing).

I don't know for sure. I think I modelled that after the `update_field` case that you did, but that situation is a bit different.

abergeron replied:
If this field is missing the render will crash.

`update_field` is different because it is either a string or the boolean value `false` to disable the functionality.

gyom replied:
When you say that the "render will crash", do you mean that Ansible will catch that error, or that it will propagate elsewhere and cause problems in Clockwork?

abergeron replied:
Ansible will fail in that case.

abergeron commented on templates/clockwork.toml.j2:
```diff
@@ -34,4 +34,20 @@ allocations="*"
+sacct_enabled=false
+{% endif %}
+sacct_path="{{ cluster.sacct_path }}"
+sacct_ssh_key_filename="{{ cluster.sacct_ssh_key_filename }}"
```
There is already an assumption that there is a single keyfile for all the clusters so it could be a separate global config.

It doesn't matter that much, but it would avoid repetition.

abergeron reviewed (commented):
Looks much better now.

abergeron commented:
> > I'm wondering why you need to add those fields to the clockwork config.
> > Normally this kind of thing should be handled by the remote data script like it is currently done for the sinfo files.
> > If you want to rework this, I'm ok with it.
> 
> Is this request still relevant given the discussion yesterday and the changes today?

No it is no longer relevant.

gyom commented:
> I'm wondering why you need to add those fields to the clockwork config.
> 
> Normally this kind of thing should be handled by the remote data script like it is currently done for the sinfo files.
> 
> If you want to rework this, I'm ok with it.

I've addressed this in a separate conversation (not on github). Waiting for this PR to be merged unless you have other comments to make.

gyom committed:
removed IF for `sacct_enabled`
